panic while editing multibuffer

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I was making an edit in multibuffer across a large project.

### Environment

Zed: v0.168.2 (Zed Preview)
OS: macOS 15.2.0
Memory: 64 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-01-06T13:41:45.988005-08:00 [ERROR] {
  "thread": "main",
  "payload": "offset 970 is out of range, max allowed is 68",
  "location_data": {
    "file": "crates/text/src/text.rs",
    "line": 2934
  },
  "backtrace": [
    "backtrace::capture::Backtrace::create::hbd8ec261cf178447+8711596",
    "backtrace::capture::Backtrace::new::h4d34e6163ddbe424+8711408",
    "zed::reliability::init_panic_hook::{{closure}}::he57b61f4ac676507+76616956",
    "std::panicking::rust_panic_with_hook::h09e8a656f11e82b2+58793924",
    "std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c+58791020",
    "std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2+58783780",
    "_rust_begin_unwind+58790208",
    "core::panicking::panic_fmt::ha4b80a05b9fff47a+84457536",
    "language::buffer::BufferSnapshot::text_object_ranges::h6d1b6d09522b62ca+66266812",
    "vim::object::text_object::h0132b7d503829dd7+66627560",
    "vim::object::<impl vim::Vim>::object::h1739b725b14be568+66610584",
    "editor::Editor::register_action::{{closure}}::{{closure}}::hf54aadec01680aef+66210472",
    "gpui::window::WindowContext::dispatch_action_on_node::hb5a0b2cbf5323d2a+30464292",
    "gpui::window::WindowContext::dispatch_event::h07447dd7cb887d77+30526204",
    "gpui::window::Window::new::{{closure}}::h0c6e3dd85d1883f6+30398448",
    "gpui::platform::mac::window::handle_key_event::h279b3de0c9855315+30252600",
    "<gpui::platform::mac::platform::MacPlatform as gpui::platform::Platform>::run::hb57b3f8c2257fc2a+30182444",
    "gpui::app::App::run::hdfff691e571da078+75271920",
    "zed::main::ha5a26be7873ebcad+76661372",
    "std::sys::backtrace::__rust_begin_short_backtrace::h2128df180c34f73c+76531444",
    "std::rt::lang_start::{{closure}}::h73a8a9c91817e8c3+76898768",
    "std::rt::lang_start_internal::hdd117cb81a316264+58725560",
    "_main+76673360"
  ],
  "app_version": "0.168.2",
  "release_channel": "preview",
  "target": "aarch64-apple-darwin",
  "os_name": "macOS",
  "os_version": "15.2.0",
  "architecture": "aarch64",
  "panicked_on": 1736199705987,
  "system_id": "620baf89-3bab-43d0-a0bd-b6dcd0cc722b",
  "installation_id": "796d8e13-af5e-4eb9-a027-92cf76869aca",
  "session_id": "3a7c7f56-2814-46a1-a009-109b6ae8533e"
}
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

<details><summary>symbolicated crash</summary>

```
Abort trap: 6 (rust panic) on thread 10237961 (main com.apple.main-thread)
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/raw_vec.rs:160:       alloc::raw_vec::RawVec$LT$T$C$A$GT$::with_capacity_in::h2ca054225068858c
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs:699:       alloc::vec::Vec$LT$T$C$A$GT$::with_capacity_in::h781d6e7520b87800
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs:481:       alloc::vec::Vec$LT$T$GT$::with_capacity::hc0fabb3ce1f99b4c
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/spec_from_iter_nested.rs:54:_$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..spec_from_iter_nested..SpecFromIterNested$LT$T$C$I$GT$$GT$::from_iter::hc7fd35d24e89c19e
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/spec_from_iter.rs:33:     _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..spec_from_iter..SpecFromIter$LT$T$C$I$GT$$GT$::from_iter::hbfd109cab8d435df
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs:2977:      _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..iter..traits..collect..FromIterator$LT$T$GT$$GT$::from_iter::h258d0b1e13778f7a
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/traits/iterator.rs:2005:  core::iter::traits::iterator::Iterator::collect::h2e5e795df1a6ac3d
crates/language/src/buffer.rs:3380:     language::buffer::BufferSnapshot::text_object_ranges::h6d1b6d09522b62ca
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/adapters/filter_map.rs:24: core::iter::adapters::filter_map::FilterMap$LT$I$C$F$GT$::new::hf1dc0ad3bb433631
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/traits/iterator.rs:935:   core::iter::traits::iterator::Iterator::filter_map::h23c2521511313718
crates/vim/src/object.rs:520:   vim::object::text_object::h0132b7d503829dd7
crates/vim/src/visual.rs:353:   vim::visual::_$LT$impl$u20$vim..Vim$GT$::visual_object::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h2162361d9b429857
crates/editor/src/selections_collection.rs:685: editor::selections_collection::MutableSelectionsCollection::move_with::_$u7b$$u7b$closure$u7d$$u7d$::h41d4c51c26f0ffc5
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/adapters/map.rs:96:       core::iter::adapters::map::map_try_fold::_$u7b$$u7b$closure$u7d$$u7d$::h5aee2fc52a493bac
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/into_iter.rs:345: _$LT$alloc..vec..into_iter..IntoIter$LT$T$C$A$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::try_fold::h083a6433c206c706
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/adapters/map.rs:122:      _$LT$core..iter..adapters..map..Map$LT$I$C$F$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::try_fold::hd64f2cf695cfbb18
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/in_place_collect.rs:379:  _$LT$I$u20$as$u20$alloc..vec..in_place_collect..SpecInPlaceCollect$LT$T$C$I$GT$$GT$::collect_in_place::h0f1a333cfcad49bd
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/in_place_collect.rs:270:  alloc::vec::in_place_collect::from_iter_in_place::haed9e65d5b9fc6ca
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/in_place_collect.rs:246:  alloc::vec::in_place_collect::_$LT$impl$u20$alloc..vec..spec_from_iter..SpecFromIter$LT$T$C$I$GT$$u20$for$u20$alloc..vec..Vec$LT$T$GT$$GT$::from_iter::h49e5d4a3c558e9c6
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/vec/mod.rs:2977:      _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..iter..traits..collect..FromIterator$LT$T$GT$$GT$::from_iter::h2e3c8b23b4bde3aa
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/iter/traits/iterator.rs:2005:  core::iter::traits::iterator::Iterator::collect::h999de527c2f7fd4b
crates/editor/src/selections_collection.rs:691: editor::selections_collection::MutableSelectionsCollection::move_with::h602b3de0ccdb9593
crates/vim/src/visual.rs:338:   vim::visual::_$LT$impl$u20$vim..Vim$GT$::visual_object::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h82e3892c83303888
crates/editor/src/selections_collection.rs:405: editor::selections_collection::SelectionsCollection::change_with::hd8eac56405abb622
crates/editor/src/editor.rs:2008:       editor::Editor::change_selections_inner::h694a304f12aeaac3
crates/editor/src/editor.rs:1995:       editor::Editor::change_selections::h04209441f4157df6
crates/vim/src/visual.rs:337:   vim::visual::_$LT$impl$u20$vim..Vim$GT$::visual_object::_$u7b$$u7b$closure$u7d$$u7d$::h2c38b1b998fcb078
crates/vim/src/vim.rs:807:      vim::Vim::update_editor::_$u7b$$u7b$closure$u7d$$u7d$::h66ba3ab1ec17bda5
crates/gpui/src/window.rs:3959: _$LT$gpui..window..WindowContext$u20$as$u20$gpui..VisualContext$GT$::update_view::h056ca42d75b14471
crates/gpui/src/window.rs:4618: _$LT$gpui..window..ViewContext$LT$V$GT$$u20$as$u20$gpui..VisualContext$GT$::update_view::h1a69c44ff98e3f84
crates/gpui/src/view.rs:77:     gpui::view::View$LT$V$GT$::update::h356ea5422f7b04e0
crates/vim/src/vim.rs:807:      vim::Vim::update_editor::h0ce0d89695597377
crates/vim/src/visual.rs:336:   vim::visual::_$LT$impl$u20$vim..Vim$GT$::visual_object::h8e4b255abf15cc56
crates/vim/src/object.rs:145:   vim::object::_$LT$impl$u20$vim..Vim$GT$::object::h1739b725b14be568
crates/gpui/src/window.rs:3960: _$LT$gpui..window..WindowContext$u20$as$u20$gpui..VisualContext$GT$::update_view::h682ff804de1b0489
crates/gpui/src/view.rs:77:     gpui::view::View$LT$V$GT$::update::h36286c723067fed4
crates/gpui/src/view.rs:193:    gpui::view::WeakView$LT$V$GT$::update::h0d6c5721613aed75
crates/gpui/src/window.rs:4540: gpui::window::ViewContext$LT$V$GT$::listener::_$u7b$$u7b$closure$u7d$$u7d$::hdb4a7671adac6836
crates/editor/src/editor.rs:12900:      editor::Editor::register_action::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hf54aadec01680aef
crates/gpui/src/window.rs:3995: _$LT$gpui..window..WindowContext$u20$as$u20$core..ops..deref..Deref$GT$::deref::h647cbad00a96ffa5
crates/gpui/src/window.rs:3568: gpui::window::WindowContext::dispatch_action_on_node::hb5a0b2cbf5323d2a
crates/gpui/src/window.rs:3995: _$LT$gpui..window..WindowContext$u20$as$u20$core..ops..deref..Deref$GT$::deref::h647cbad00a96ffa5
crates/gpui/src/window.rs:3354: gpui::window::WindowContext::dispatch_key_event::hdde53231d74746f7
crates/gpui/src/window.rs:3183: gpui::window::WindowContext::dispatch_event::h07447dd7cb887d77
crates/gpui/src/window.rs:776:  gpui::window::Window::new::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h5593bd2e656eae78
crates/gpui/src/app.rs:1504:    _$LT$gpui..app..AppContext$u20$as$u20$gpui..Context$GT$::update_window::_$u7b$$u7b$closure$u7d$$u7d$::h65f3148e2b17dbb6
crates/gpui/src/app.rs:418:     gpui::app::AppContext::update::h334e5e6e8a124594
crates/gpui/src/app.rs:1495:    _$LT$gpui..app..AppContext$u20$as$u20$gpui..Context$GT$::update_window::he6c341736a031b63
crates/gpui/src/app/async_context.rs:91:        _$LT$gpui..app..async_context..AsyncAppContext$u20$as$u20$gpui..Context$GT$::update_window::h8b5aa75cf9253cc5
crates/gpui/src/window.rs:4846: gpui::window::AnyWindowHandle::update::h99041e654c39114c
crates/gpui/src/window.rs:776:  gpui::window::Window::new::_$u7b$$u7b$closure$u7d$$u7d$::h0c6e3dd85d1883f6
crates/gpui/src/platform/mac/window.rs:1292:    gpui::platform::mac::window::handle_key_event::h279b3de0c9855315
AppKit  -[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:]
AppKit  -[NSWindow(NSEventRouting) sendEvent:]
AppKit  -[NSApplication(NSEventRouting) sendEvent:]
AppKit  -[NSApplication _handleEvent:]
AppKit  -[NSApplication run]
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/sync/atomic.rs:3313:   core::sync::atomic::atomic_load::h77b5fa95820e3824
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/sync/atomic.rs:2421:   core::sync::atomic::AtomicUsize::load::h15b754d867e166fa
/Users/administrator/.cargo/registry/src/index.crates.io-6f17d22bba15001f/objc-0.2.7/src/macros.rs:54:    _$LT$$BP$mut$u20$objc..runtime..Object$u20$as$u20$cocoa_foundation..foundation..NSAutoreleasePool$GT$::drain::register_sel::h98539e22d846bff9
/Users/administrator/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cocoa-foundation-0.2.0/src/foundation.rs:192:   _$LT$$BP$mut$u20$objc..runtime..Object$u20$as$u20$cocoa_foundation..foundation..NSAutoreleasePool$GT$::drain::h4502b15bab1b6b2c
crates/gpui/src/platform/mac/platform.rs:454:   _$LT$gpui..platform..mac..platform..MacPlatform$u20$as$u20$gpui..platform..Platform$GT$::run::hb57b3f8c2257fc2a
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/rc.rs:3455:   alloc::rc::RcInnerPtr::dec_strong::h4eadbbab1bd8a48e
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/alloc/src/rc.rs:2219:   _$LT$alloc..rc..Rc$LT$T$C$A$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h3ad13cea38779d2f
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:        core::ptr::drop_in_place$LT$alloc..rc..Rc$LT$dyn$u20$gpui..platform..Platform$GT$$GT$::h8eb2b07df3335102
crates/gpui/src/app.rs:170:     gpui::app::App::run::hdfff691e571da078
/rust/deps/hashbrown-0.14.5/src/raw/mod.rs:2823:        hashbrown::raw::RawTableInner::is_empty_singleton::h0062a2217e06ab9a
/rust/deps/hashbrown-0.14.5/src/raw/mod.rs:2429:        hashbrown::raw::RawTableInner::drop_inner_table::h79bdfbae57929046
/rust/deps/hashbrown-0.14.5/src/raw/mod.rs:3673:        _$LT$hashbrown..raw..RawTable$LT$T$C$A$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hf626b4bb245a3872
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:        core::ptr::drop_in_place$LT$hashbrown..raw..RawTable$LT$$LP$std..io..error..ErrorKind$C$alloc..vec..Vec$LT$$RF$std..path..Path$GT$$RP$$GT$$GT$::h837d9aef36844b9d
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:        core::ptr::drop_in_place$LT$hashbrown..map..HashMap$LT$std..io..error..ErrorKind$C$alloc..vec..Vec$LT$$RF$std..path..Path$GT$$C$std..hash..random..RandomState$GT$$GT$::h95b0746f6d62543c
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ptr/mod.rs:542:        core::ptr::drop_in_place$LT$std..collections..hash..map..HashMap$LT$std..io..error..ErrorKind$C$alloc..vec..Vec$LT$$RF$std..path..Path$GT$$GT$$GT$::hb00d2dcf6b3ce816
crates/zed/src/main.rs:612:     zed::main::ha5a26be7873ebcad
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/rt.rs:162:      std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h73a8a9c91817e8c3
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/result.rs:1433:        core::result::Result$LT$T$C$E$GT$::unwrap_or::hf4ab4d10c9f1b6dd
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/rt.rs:141:      std::rt::lang_start_internal::_$u7b$$u7b$closure$u7d$$u7d$::hf216622dc2c733e3
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:557:       std::panicking::try::do_call::haa691957db1dd55f
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:521:       std::panicking::try::ha0c1a49b9fabc98e
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panic.rs:350:   std::panic::catch_unwind::h68ad032c646bb0ab
/rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/rt.rs:141:      std::rt::lang_start_internal::hdd117cb81a316264
main (in Zed-0.168.2-aarch64-apple-darwin.dwarf) + 52
dyld    start
```
</details>

--- a/crates/vim/src/object.rs
+++ b/crates/vim/src/object.rs
@@ -514,13 +514,15 @@ fn text_object(
 
     let excerpt = snapshot.excerpt_containing(offset..offset)?;
     let buffer = excerpt.buffer();
+    let offset = excerpt.map_offset_to_buffer(offset);
 
     let mut matches: Vec<Range<usize>> = buffer
         .text_object_ranges(offset..offset, TreeSitterOptions::default())
         .filter_map(|(r, m)| if m == target { Some(r) } else { None })
         .collect();
     matches.sort_by_key(|r| (r.end - r.start));
-    if let Some(range) = matches.first() {
+    if let Some(buffer_range) = matches.first() {
+        let range = excerpt.map_range_from_buffer(buffer_range.clone());
         return Some(range.start.to_display_point(map)..range.end.to_display_point(map));
     }
 
@@ -537,12 +539,14 @@ fn text_object(
         .filter_map(|(r, m)| if m == target { Some(r) } else { None })
         .collect();
     matches.sort_by_key(|r| r.start);
-    if let Some(range) = matches.first() {
-        if !range.is_empty() {
+    if let Some(buffer_range) = matches.first() {
+        if !buffer_range.is_empty() {
+            let range = excerpt.map_range_from_buffer(buffer_range.clone());
             return Some(range.start.to_display_point(map)..range.end.to_display_point(map));
         }
     }
-    return Some(around_range.start.to_display_point(map)..around_range.end.to_display_point(map));
+    let buffer_range = excerpt.map_range_from_buffer(around_range.clone());
+    return Some(buffer_range.start.to_display_point(map)..buffer_range.end.to_display_point(map));
 }
 
 fn argument(

